source/text/Font.cpp

```diff
@@ -14,573 +14,823 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
-		glyphW,    0.f, 1.f, 0.f,
-		glyphW, glyphH, 1.f, 1.f
-	};
-	glBufferData(GL_ARRAY_BUFFER, sizeof(vertices), vertices, GL_STATIC_DRAW);
+	UpdateFont();
+}
+
+
+
+void Font::UpdateFont() const
+{
+	if(pixelSize <= 0)
+		return;
 	
-	// Connect the xy to the "vert" attribute of the vertex shader.
-	constexpr auto stride = 4 * sizeof(GLfloat);
-	glEnableVertexAttribArray(shader.Attrib("vert"));
-	glVertexAttribPointer(shader.Attrib("vert"), 2, GL_FLOAT, GL_FALSE, stride, nullptr);
+	// Get font descriptions.
+	PangoFontDescription *fontDesc = pango_font_description_from_string(drawingSettings.description.c_str());
 	
-	glEnableVertexAttribArray(shader.Attrib("corner"));
-	glVertexAttribPointer(shader.Attrib("corner"), 2, GL_FLOAT, GL_FALSE,
-		stride, reinterpret_cast<const GLvoid *>(2 * sizeof(GLfloat)));
+	// Set the pixel size.
+	const int fontSize = ViewFromTextFloorY(pixelSize) * PANGO_SCALE;
+	pango_font_description_set_absolute_size(fontDesc, fontSize);
 	
-	glBindBuffer(GL_ARRAY_BUFFER, 0);
-	glBindVertexArray(0);
+	// Update the context.
+	PangoLanguage *lang = pango_language_from_string(drawingSettings.language.c_str());
+	pango_context_set_language(context, lang);
 	
-	// We must update the screen size next time we draw.
-	screenWidth = 0;
-	screenHeight = 0;
+	// Update the layout.
+	pango_layout_set_font_description(pangoLayout, fontDesc);
 	
-	colorI = shader.Uniform("color");
-	scaleI = shader.Uniform("scale");
-	glyphI = shader.Uniform("glyph");
-	aspectI = shader.Uniform("aspect");
-	positionI = shader.Uniform("position");
+	// Update layout parameters.
+	PangoFontMetrics *metrics = pango_context_get_metrics(context, fontDesc, lang);
+	const int ascent = pango_font_metrics_get_ascent(metrics);
+	const int descent = pango_font_metrics_get_descent(metrics);
+	viewFontHeight = PixelFromPangoCeil(ascent + descent);
+	if (drawingSettings.lineHeightScale >= 0)
+		viewDefaultLineHeight = viewFontHeight * drawingSettings.lineHeightScale;
+	else
+		viewDefaultLineHeight = 0;
+	if (drawingSettings.paragraphBreakScale >= 0)
+		viewDefaultParagraphBreak = viewFontHeight * drawingSettings.paragraphBreakScale;
+	else
+		viewDefaultParagraphBreak = 0;
+	
+	// Clean up.
+	pango_font_metrics_unref(metrics);
+	pango_font_description_free(fontDesc);
+	cache.Clear();
+	
+	// Tab Stop
+	space = ViewWidth(DisplayText(" ", {}));
+	const int tabSize = 4 * space * PANGO_SCALE;
+	PangoTabArray *tb = pango_tab_array_new(TOTAL_TAB_STOPS, FALSE);
+	for(int i = 0; i < TOTAL_TAB_STOPS; ++i)
+		pango_tab_array_set_tab(tb, i, PANGO_TAB_LEFT, i * tabSize);
+	pango_layout_set_tabs(pangoLayout, tb);
+	pango_tab_array_free(tb);
 }
 
 
 
-int Font::WidthRawString(const char *str, char after) const noexcept
+// Replaces straight quotation marks with curly ones,
+// and escapes "&" except for minimum necessaries because a pilot name may
+// contain "&" and that representation should be the same as the old version.
+string Font::ReplaceCharacters(const string &str)
```

Comment:
This static method will be a local function because it's private.
I plan to create two functions in FontUtilities to escape/revert special characters.